python/fnllm/fnllm/openai/factories/chat.py

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2024 Microsoft Corporation.
+
+"""Factory functions for creating OpenAI LLMs."""
```

Comment:
I think the name Factory can be a little misleading since it is more a collection of creation methods. I would suggest we rename, or we align it to follow a factory pattern

Reply:
These are the main factory methods for creating OpenAI LLMs though - `create_openai_chat_llm`, `create_openai_embeddings_llm`, and `create_openai_client` are intended to be used by consumers of the library.

The term `factory`, in my mind, implies a function that constructs objects, whether that's a module-level function or a static class member. You may be thinking of `builders`, which utilize fluent APIs to chain together the parameters of the thing you're building (e.g. `ConnectionstringBuilder().name(username).password(password).build()` 

In what way do these seem to deviate from the factory pattern to you?

Reply:
I usually think of factories as a magic function/object that creates different objects for me, for example create_chat_llm(Client, other params) and that returns me a protocol/interface type that I can call without caring if it's openai or other... at some point I was thinking of doing a functional kind of factory that actually would return a function something like:

```python
config = OpenAIConfig(...) # or AzureConfig or AnyConfig()

chat_completion = create_chat_completion(config, limiters, etc...) ) # this instantiates the client based on the config and returns a function that you can call for completions

response = chat_completion("some prompt", history=["blabla"])

```

Reply:
Nope, I didn't mean that we should follow a builder pattern, but a proper Class Factory or a Factory Method in which you just have a method that instantiates the proper interface or object depending on the parameters you pass. That way you don't actually care about how to or where to instantiate something.

Reply:
> for me, for example create_chat_llm(Client, other params) and that returns me a protocol/interface type that I can call without caring if it's openai or other

> but a proper Class Factory or a Factory Method in which you just have a method that instantiates the proper interface or object depending on the parameters you pas

Right, so I think these functions essentially satisfy these conditions, and are okay to leave as-is. We can explore more generic factory functions as we roll those other implementations in. These functions are under `fnllm.openai.factories`, so I don't think it's unexpected that they build openai-specific objects.